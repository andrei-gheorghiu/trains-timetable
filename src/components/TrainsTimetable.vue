<template>
  <div class="container">
    <div class="form-wrapper">
      <v-select
        v-model="currentStation"
        :get-option-label="getStationLabel"
        :options="stations"
      ></v-select>
      <v-select
        v-model="currentDirection"
        :reduce="reduceDirection"
        :options="directions"
      ></v-select>
    </div>
    <div class="card">
      <table>
        <thead>
          <tr>
            <th>No.</th>
            <th>Destination</th>
            <th>Direction</th>
            <th>Due in</th>
          </tr>
        </thead>
        <tbody v-if="filteredStationData.length">
          <tr v-for="(record, key) in filteredStationData" :key="key">
            <td v-text="key + 1" />
            <td v-text="record.destination" />
            <td v-text="record.direction" />
            <td v-text="humanizeMinutes(+record.duein)" />
          </tr>
        </tbody>
        <tbody v-else>
          <tr>
            <td colspan="4">
              <div class="no-data">
                {{
                  currentStation
                    ? "No trains"
                    : "You haven't selected a station."
                }}
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="card d-flex space-between">
      <div class="small">
        Last updated:
        <code>{{ currentStation ? counter.val.toFixed(1) : "--" }}s</code> ago.
      </div>
      <div class="small">
        Next update: in
        <code>{{ currentStation ? (20 - counter.val).toFixed(1) : "--" }}s</code
        >.
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  onMounted,
  reactive,
  toRefs,
  watch,
} from "vue";
import vSelect from "vue-select";
import { useTimetable } from "@/store";
import { Station, StationData, Train } from "@/types";
import { baseDirections } from "@/util/helpers";
import gsap, { Power0 } from "gsap";
import humanize from "humanize-duration";

interface TrainsTimetableState {
  trains: Train[];
  stations: Station[];
  currentStation: Station | null;
  stationData: StationData[];
  filteredStationData: StationData[];
  directions: { label: string; value: string }[];
  currentDirection: string;
  counter: { val: number };
}

interface GenericOption {
  label: string;
  value: unknown;
}

export default defineComponent({
  name: "TrainsTimetable",
  components: {
    vSelect,
  },
  setup() {
    const tt = useTimetable();
    onMounted(async () => {
      await tt.fetchStations();
      tt.currentStation =
        tt.stations.find((st) => st.stationDesc === "Dublin Pearse") || null;
    });
    const state: TrainsTimetableState = reactive({
      trains: computed(() => tt.trains),
      stations: computed(() => tt.stations),
      currentStation: computed({
        get(): Station | null {
          return tt.currentStation;
        },
        set(val: Station | null) {
          tt.currentStation = val;
        },
      }),
      stationData: computed(() => tt.stationData),
      filteredStationData: computed(() =>
        tt.stationData
          .filter(
            (sd) =>
              !state.currentDirection || state.currentDirection === sd.direction
          )
          .sort((a, b) => (+a.duein > +b.duein ? 1 : -1))
      ),
      directions: computed(() =>
        [
          ...new Set([
            ...baseDirections,
            ...tt.stationData.map((t) => t.direction),
          ]),
        ].map((label) => ({ label, value: label }))
      ),
      currentDirection: "Southbound",
      counter: { val: 0 },
    });
    let tween: gsap.core.Tween;
    const iteration = () => {
      tween = gsap.fromTo(
        state.counter,
        { val: 0, decimals: 0 },
        {
          val: 20,
          duration: 20,
          ease: Power0.easeNone,
          onComplete: updateStation,
        }
      );
    };
    const updateStation = () => {
      tween?.kill();
      if (tt.currentStation?.stationCode) {
        tt.fetchStationByCode({
          StationCode: tt.currentStation.stationCode,
        }).then(iteration);
      } else {
        tt.stationData = [];
      }
    };
    const getStationLabel = (item: Station) => item.stationDesc;
    const humanizeMinutes = (n: number) => humanize(n * 6e4);
    const reduceDirection = (item: GenericOption) => item.value;
    watch(() => state.currentStation, updateStation);
    return {
      humanizeMinutes,
      reduceDirection,
      getStationLabel,
      ...toRefs(state),
    };
  },
});
</script>

<style scoped lang="scss">
.container {
  margin: 0 auto;
  max-width: 1000px;
}
.form-wrapper {
  display: flex;
  align-items: center;
  > * {
    flex: 1 0 0;
    margin: 0 0.5rem;
    &:first-child {
      margin-left: 0;
    }
    &:last-child {
      margin-right: 0;
    }
  }
}
.small {
  font-family: monospace;
  font-size: 1rem;
  code {
    color: #ffd114;
  }
}
table {
  width: 100%;
  th {
    text-align: left;
    border-bottom: 1px solid rgba(255, 255, 255, 0.42);
    &:first-child {
      width: 30px;
    }
  }
  .no-data {
    padding: 5rem;
    display: flex;
    justify-content: center;
    font-style: italic;
  }
  td,
  th {
    padding: 3px 7px;
  }
}
.d-flex {
  display: flex;
}
.space-between {
  justify-content: space-between;
}
</style>
