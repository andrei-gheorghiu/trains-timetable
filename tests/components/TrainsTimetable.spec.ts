import { shallowMount } from "@vue/test-utils";
import TrainsTimetable from "@/components/TrainsTimetable.vue";
import { createTestingPinia } from "@pinia/testing";

describe("TrainsTimetable.vue", () => {
  let consoleSpy: any;
  let wrapper: any;

  beforeEach(() => {
    consoleSpy = jest.spyOn(console, "error");
    wrapper = shallowMount(TrainsTimetable, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
  });

  afterEach(() => {
    expect(consoleSpy).not.toHaveBeenCalled();
  });

  it("should render without errors", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
