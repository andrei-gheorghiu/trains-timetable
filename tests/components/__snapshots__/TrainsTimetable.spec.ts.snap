// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrainsTimetable.vue should render without errors 1`] = `
"<div class=\\"container\\">
  <div class=\\"form-wrapper\\">
    <v-select-stub components=\\"[object Object]\\" options=\\"\\" disabled=\\"false\\" clearable=\\"true\\" deselectfromdropdown=\\"false\\" searchable=\\"true\\" multiple=\\"false\\" placeholder=\\"\\" transition=\\"vs__fade\\" clearsearchonselect=\\"true\\" closeonselect=\\"true\\" label=\\"label\\" autocomplete=\\"off\\" reduce=\\"function(e){return e}\\" selectable=\\"function(e){return!0}\\" getoptionlabel=\\"function (item) {
      /* istanbul ignore next */
      cov_17qaly50fu().f[18]++;
      cov_17qaly50fu().s[40]++;
      return item.stationDesc;
    }\\" getoptionkey=\\"function(e){if(&quot;object&quot;!==g()(e))return e;try{return e.hasOwnProperty(&quot;id&quot;)?e.id:T(e)}catch(t){return console.warn(&quot;[vue-select warn]: Could not stringify this option to generate unique key. Please provide'getOptionKey' prop to return a unique key for each option.\\\\nhttps://vue-select.org/api/props.html#getoptionkey&quot;,e,t)}}\\" taggable=\\"false\\" pushtags=\\"false\\" filterable=\\"true\\" filterby=\\"function(e,t,o){return(t||&quot;&quot;).toLocaleLowerCase().indexOf(o.toLocaleLowerCase())>-1}\\" filter=\\"function(e,t){var o=this;return e.filter((function(e){var n=o.getOptionLabel(e);return&quot;number&quot;==typeof n&amp;&amp;(n=n.toString()),o.filterBy(e,n,t)}))}\\" createoption=\\"function(e){return&quot;object&quot;===g()(this.optionList[0])?j()({},this.label,e):e}\\" resetonoptionschange=\\"false\\" clearsearchonblur=\\"function(e){var t=e.clearSearchOnSelect,o=e.multiple;return t&amp;&amp;!o}\\" nodrop=\\"false\\" dir=\\"auto\\" selectontab=\\"false\\" selectonkeycodes=\\"13\\" searchinputqueryselector=\\"[type=search]\\" mapkeydown=\\"function(e,t){return e}\\" appendtobody=\\"false\\" calculateposition=\\"function(e,t,o){var n=o.width,i=o.top,r=o.left;e.style.top=i,e.style.left=r,e.style.width=n}\\" dropdownshouldopen=\\"function(e){var t=e.noDrop,o=e.open,n=e.mutableLoading;return!t&amp;&amp;(o&amp;&amp;!n)}\\" uid=\\"1\\"></v-select-stub>
    <v-select-stub modelvalue=\\"Southbound\\" components=\\"[object Object]\\" options=\\"\\" disabled=\\"false\\" clearable=\\"true\\" deselectfromdropdown=\\"false\\" searchable=\\"true\\" multiple=\\"false\\" placeholder=\\"\\" transition=\\"vs__fade\\" clearsearchonselect=\\"true\\" closeonselect=\\"true\\" label=\\"label\\" autocomplete=\\"off\\" reduce=\\"function (item) {
      /* istanbul ignore next */
      cov_17qaly50fu().f[20]++;
      cov_17qaly50fu().s[44]++;
      return item.value;
    }\\" selectable=\\"function(e){return!0}\\" getoptionlabel=\\"function(e){return&quot;object&quot;===g()(e)?e.hasOwnProperty(this.label)?e[this.label]:console.warn('[vue-select warn]: Label key &quot;option.'.concat(this.label,'&quot; does not')+&quot; exist in options object &quot;.concat(JSON.stringify(e),&quot;.\\\\n&quot;)+&quot;https://vue-select.org/api/props.html#getoptionlabel&quot;):e}\\" getoptionkey=\\"function(e){if(&quot;object&quot;!==g()(e))return e;try{return e.hasOwnProperty(&quot;id&quot;)?e.id:T(e)}catch(t){return console.warn(&quot;[vue-select warn]: Could not stringify this option to generate unique key. Please provide'getOptionKey' prop to return a unique key for each option.\\\\nhttps://vue-select.org/api/props.html#getoptionkey&quot;,e,t)}}\\" taggable=\\"false\\" pushtags=\\"false\\" filterable=\\"true\\" filterby=\\"function(e,t,o){return(t||&quot;&quot;).toLocaleLowerCase().indexOf(o.toLocaleLowerCase())>-1}\\" filter=\\"function(e,t){var o=this;return e.filter((function(e){var n=o.getOptionLabel(e);return&quot;number&quot;==typeof n&amp;&amp;(n=n.toString()),o.filterBy(e,n,t)}))}\\" createoption=\\"function(e){return&quot;object&quot;===g()(this.optionList[0])?j()({},this.label,e):e}\\" resetonoptionschange=\\"false\\" clearsearchonblur=\\"function(e){var t=e.clearSearchOnSelect,o=e.multiple;return t&amp;&amp;!o}\\" nodrop=\\"false\\" dir=\\"auto\\" selectontab=\\"false\\" selectonkeycodes=\\"13\\" searchinputqueryselector=\\"[type=search]\\" mapkeydown=\\"function(e,t){return e}\\" appendtobody=\\"false\\" calculateposition=\\"function(e,t,o){var n=o.width,i=o.top,r=o.left;e.style.top=i,e.style.left=r,e.style.width=n}\\" dropdownshouldopen=\\"function(e){var t=e.noDrop,o=e.open,n=e.mutableLoading;return!t&amp;&amp;(o&amp;&amp;!n)}\\" uid=\\"2\\"></v-select-stub>
  </div>
  <div class=\\"card table-wrapper\\">
    <table>
      <thead>
        <tr>
          <th>No.</th>
          <th>Destination</th>
          <th>Direction</th>
          <th>Due in</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td colspan=\\"4\\">
            <div class=\\"no-data\\">You haven't selected a station.</div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class=\\"card d-flex space-between\\">
    <div class=\\"small\\"> Last updated: <code>--s</code> ago. </div>
    <div class=\\"small\\"> Next update: in <code>--s</code>. </div>
  </div>
</div>"
`;
